composinator/convert.go

```diff
@@ -0,0 +1,517 @@
+package composinator
+
+import (
+	"fmt"
+	"net/http"
+	"path"
+	"strings"
+
+	"encoding/json"
+
+	"github.com/pkg/errors"
+	v3 "github.com/rancher/go-rancher/v3"
+	"github.com/rancher/rancher-compose-executor/config"
+	"github.com/rancher/rancher-compose-executor/utils"
+	"github.com/rancher/rancher-compose-executor/yaml"
+	yml "gopkg.in/yaml.v2"
+)
+
+const (
+	managedNetwork         = "managed"
+	containerNetwork       = "container"
+	labelSelectorContainer = "io.rancher.service.selector.container"
+	labelServiceGlobal     = "io.rancher.scheduler.global"
+	virtualMachineKind     = "virtualMachine"
+	hashLabel              = "io.rancher.service.hash"
+	blkioWeight            = "weight"
+	blkioReadIops          = "readIops"
+	blkioReadBps           = "readBps"
+	blkioWriteIops         = "writeIops"
+	blkioWriteBps          = "writeBps"
+)
+
+func convert(w http.ResponseWriter, client *v3.RancherClient, input convertOptions) {
+	services, volumeTemplates, certMap, serviceMap, containerMap, secretMap, err := prepareExportedData(client, input.StackID)
+	if err != nil {
+		http.Error(w, "can't obtain exported data", http.StatusInternalServerError)
+	}
+	dockerCompose, rancherCompose, compose, err := createComposeData(services, volumeTemplates, certMap, serviceMap, containerMap, secretMap, input.Format)
+	if err != nil {
+		http.Error(w, "can't create compose file", http.StatusInternalServerError)
+	}
+	result := map[string]string{}
+	if input.Format == "combined" {
+		result["compose"] = compose
+	} else {
+		result["dockerCompose"] = dockerCompose
+		result["rancherCompose"] = rancherCompose
+	}
+	w.WriteHeader(http.StatusOK)
+	data, err := json.Marshal(result)
+	if err != nil {
+		http.Error(w, "can't marshall result", http.StatusInternalServerError)
+	}
+	fmt.Fprint(w, data)
+}
+
+func prepareExportedData(client *v3.RancherClient, stackID string) ([]*v3.Service, map[string]v3.VolumeTemplate, map[string]v3.Certificate, map[string]v3.Service, map[string]v3.Instance, map[string]v3.Secret, error) {
+	volumeTemplates, err := client.VolumeTemplate.List(&v3.ListOpts{
+		Filters: map[string]interface{}{
+			"stackId":      stackID,
+			"removed_null": "true",
+			"limit":        "-1",
+		},
+	})
+	if err != nil {
+		return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't list volumeTemplates")
+	}
+	stack, err := client.Stack.ById(stackID)
+	if err != nil {
+		return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't get stack")
+	}
+	servicesToExport := []*v3.Service{}
+	// lbcertMap is used for certificate reference by lbConfig
+	lbcertMap := map[string]v3.Certificate{}
+	// serviceMap is used for reference by portRule
+	serviceMap := map[string]v3.Service{}
+	containerMap := map[string]v3.Instance{}
```

Comment:
for containers you need to grab all containers where `stackId = current` and `serviceId = null`  If serviceId is null then it is a standalone container and should be exported as such